### ISSUE TITLE
Failing to reproduce results

### ISSUE THREAD

**Issue by Birch-san** [weixi-feng/Structured-Diffusion-Guidance#2]
Congratulations on the [arxiv submission](https://arxiv.org/abs/2212.05032)!

I tried to reproduce the results of this paper on top of [Huggingface Diffusers](https://github.com/huggingface/diffusers), based on the reference implementation provided in the [preprint](https://openreview.net/forum?id=PUIqjT4rzq7).  

I ended up implementing like so:  
Changes [to txt2img](https://github.com/Birch-san/diffusers-play/compare/5d0492cf7c2711157702824537f6904295aa6a7a...7ff06988631bce4a17eb08799e5c9ab67f226a0a)  
Changes [to diffusers](https://github.com/Birch-san/diffusers/compare/fc94c60c8373862c509e388f3f4065d98cedf589...58a5f7d3252f5c967ee2225987b5ec093d63c766)  
Some explanation in [tweet](https://twitter.com/Birchlabs/status/1602105002448805891).

In my independent implementation: structured diffusion changes the images **only slightly**, and in the 10 samples * 4 prompts that I tried, _never_ made the generations more relevant to the prompt.

structured (left) / regular (right) "two blue sheep and a red goat":  
<img width="128" src="https://user-images.githubusercontent.com/6141784/207460981-1c8f4308-9f00-4f91-8505-f1fbfb1a9bd0.png"><img width="128" src="https://user-images.githubusercontent.com/6141784/207461050-c7d480f7-2276-42dd-b898-ad01adb8ee90.png">
<img width="128" src="https://user-images.githubusercontent.com/6141784/207461435-136cda5f-86e2-4375-b67a-32bc05d1f855.png"><img width="128" src="https://user-images.githubusercontent.com/6141784/207461581-82104077-90bd-4471-97cb-0aee074ecf2d.png">
<img width="128" src="https://user-images.githubusercontent.com/6141784/207461747-4436c7b9-9dd2-4260-b97d-d4b4d8ba898f.png"><img width="128" src="https://user-images.githubusercontent.com/6141784/207461788-bc7553bf-c09c-4718-823d-45257f8e91eb.png">
<img width="128" src="https://user-images.githubusercontent.com/6141784/207461857-50181766-0aa0-4b19-b836-92368d7d3ec4.png"><img width="128" src="https://user-images.githubusercontent.com/6141784/207461897-48289b13-0d2b-4d6c-ae6d-9b21c7ec3871.png">

I attach the rest of my results:  
[A red bird and a green apple.zip](https://github.com/weixi-feng/Structured-Diffusion-Guidance/files/10222748/A.red.bird.and.a.green.apple.zip)
[A white goat standing next to two black goats.zip](https://github.com/weixi-feng/Structured-Diffusion-Guidance/files/10222749/A.white.goat.standing.next.to.two.black.goats.zip)
[two blue sheep and a red goat.zip](https://github.com/weixi-feng/Structured-Diffusion-Guidance/files/10222750/two.blue.sheep.and.a.red.goat.zip)
[Two ripe spotted bananas are sitting inside a green bowl on a gray counter.zip](https://github.com/weixi-feng/Structured-Diffusion-Guidance/files/10222751/Two.ripe.spotted.bananas.are.sitting.inside.a.green.bowl.on.a.gray.counter.zip)

Basically, I'm wondering whether:
- this is exactly the kind of difference I should expect to see (in line with the claimed 5–8% advantage)
- there's a mistake in my reproduction; better results are possible

Could you possibly read my [`attention.py`](https://github.com/Birch-san/diffusers/compare/fc94c60c8373862c509e388f3f4065d98cedf589...58a5f7d3252f5c967ee2225987b5ec093d63c766#diff-852d5806e41c100ee99c68c9e1059bfad9c0177e6f4416f0faf5a2f2a6a64067R610) and see if it looks like a reasonable interpretation of your algorithm? I changed it substantially to make it to do more work in parallel. I _think_ it should be equivalent, but did I miss something important?

Thanks in advance for any attention you can give this!

**Comment by weixi-feng** [comment]
Hi,

Thank you for your interest in our work. We also observe that a certain amount of our images are similar to the stable diffusion outputs. As is written in the manuscript, for evaluations, we discarded 20% of the most similar pairs and randomly sampled around 1,500 image pairs for comparison. So the 5-8% improvement could be discounted when considering all images. Among the winning cases in the head-to-head comparison, there are 31% for “fewer missing objects”, 14.1% for “better-matched colors”, and 54.8% for “other attributes or details”. So it is within expectation to see many images with details enhanced (like the last one in "two blue sheep and a red goat", 00054/64, 00058/68 in "a red bird and a green apple", 00026/36 in "a white goat standing..."). I also tried the banana prompt and observed detail enhancement on the "bananas" in 3/10 cases, while the rest look similar. It could be that we randomly ran into some good initialization for Fig. 4 while in general, the improvement is not significant for the banana prompt. 

For the conjunction prompt (i.e., using "and" to connect two objects), you may want to try [multiple keys and a single value](https://github.com/weixi-feng/Structured-Diffusion-Guidance/blob/a5b8188345d72304e8a6c0d6accb3f21a049da3a/scripts/txt2img_demo.py#L449). Note that the single value is not the plain encoding of the original prompt but an [aligned version](https://github.com/weixi-feng/Structured-Diffusion-Guidance/blob/a5b8188345d72304e8a6c0d6accb3f21a049da3a/scripts/txt2img_demo.py#L148) (also see eq. 5-6). This method is more likely to generate both objects simultaneously (like Fig. 1 (right) or Fig. 5 (top right)). However, as reflected in Table 2, these prompts seem quite challenging for existing T2I models, and we still expect incomplete compositions in most cases. If you can run this codebase and fix the seed to 42, you should be able to get the following results.

multiple keys, single value (right) / single key, multiple values (middle) / regular (right)
![combine_images](https://user-images.githubusercontent.com/19576665/207529164-0d3a0deb-5510-4b80-b957-e1129e857892.jpeg)


I am not entirely sure, but your implementation looks correct. You may use the following initialized noise patterns and see if they result in the same images as above. 
[init.zip](https://github.com/weixi-feng/Structured-Diffusion-Guidance/files/10225205/init.zip)


As mentioned in README, improvement is not guaranteed on every sample but rather system-level. Overall we find it hard to quantify images compositionally, and we are still working on metrics beyond human evaluation to improve the experiment section. You may download this [batch](https://drive.google.com/file/d/1VcXIXFcB3G7EoEBILi_PAMqMjF4JAbUp/view?usp=sharing) of examples to have a better understanding of the overall performance. Please let me know if this helps and if you have further questions. 

Thanks,
Weixi

**Comment by Birch-san** [comment]
thanks very much for the detailed response!  
okay, looking at the samples from your Google Drive: yeah, they seem to vary in a pretty similar way to the way my own results varied. that's heartening; maybe my reproduction is close or equivalent.

I tried to see if I could generate the same images as from your Google Drive (even by turning off structured diffusion and trying to match your vanilla results).

I used the 4 noised latents `.pt`s you provided, prompt "a red car and a white sheep", 15 steps DPM-Solver++ (2M) (k-diffusion), 16-bit Unet, 32-bit latents, 32-bit sampling, stable-diffusion 1.4, diffusers.

with structured diffusion _off_:  
<img width="128" src="https://user-images.githubusercontent.com/6141784/207712040-1b89fb9e-df61-4eaa-8d8b-e5fb95b079e7.png" title="0.pt"><img width="128" src="https://user-images.githubusercontent.com/6141784/207712061-b36e0822-7f63-42d8-b823-2237ef1f853c.png" title="1.pt"><img width="128" src="https://user-images.githubusercontent.com/6141784/207712067-8b60e5f0-5508-4932-9d76-3cb0605b4dec.png" title="2.pt"><img width="128" src="https://user-images.githubusercontent.com/6141784/207712071-56d5e65d-5c70-4fff-bb03-4d719ef9d406.png" title="3.pt">

these don't match the "baseline" samples from your Google Drive:  
<img width="128" src="https://user-images.githubusercontent.com/6141784/207713387-7e68bce9-ea93-48a5-b943-2a5dd35e4e12.png" title="00000-0-a red car and a white sheep"><img width="128" src="https://user-images.githubusercontent.com/6141784/207713392-fe89f60c-f5f1-43cc-afbc-b1ccc0a70e52.png" title="00001-0-a red car and a white sheep"><img width="128" src="https://user-images.githubusercontent.com/6141784/207713399-2bdc1527-d903-4516-9495-17b5bac88f49.png" title="00002-0-a red car and a white sheep">

err, well `3.pt` came out like `00000-0-a red car and a white sheep.png`. seems more likely to be a coincidence though.

maybe I'll have to resort to getting the reference implementation running in order to do a comparison.

thanks for mentioning "multiple keys, single value". I think I built something like that along the way, but [deleted it](https://github.com/Birch-san/diffusers-play/commit/f974c89b30f8070c20545e88b99a425479aba396) thinking I'd misunderstood. originally I had indeed written it such that it aligned all the noun-phrase embeddings onto one prompt embedding (instead of onto several). so I can undelete that and try it out.

"many attributes with details enhanced" might explain why (my implementation of) structured diffusion upgraded my bird into a photorealistic one ("A red bird and a green apple"):

standard | structured
<img width="128" src="https://user-images.githubusercontent.com/6141784/207716428-d88e5c82-a12a-4686-836e-f4c8f5488f93.png" title="00001-0-a red car and a white sheep"><img width="128" src="https://user-images.githubusercontent.com/6141784/207716455-f1153a29-b83e-4ac4-9b94-01e63948e449.png" title="00002-0-a red car and a white sheep">

I guess next step is for me to try running the reference implementation on my machine, see if I can get the same _baseline_ outputs, then see if I can get the same _structured diffusion_ outputs with my algorithm. if it is indeed an equivalent implementation: it'd mean we can enjoy improved perf (does more work in parallel and fuses a lot of multiplication) and diffusers support. the downside however is that if my implementation _is_ equivalent, then the results I got would be valid too (which didn't come close to the best results from the paper). but I haven't tried "multiple keys, single value", so that's worth exploration too.

**Comment by weixi-feng** [comment]
I forgot to mention that the prompt is "a white car and a red sheep", and the provided init noise patterns correspond to the 12 images displayed in the reply, not to any images in Google Drive.

Even with the same noise initialization, there might be some other randomness that causes slight differences between the provided images and your generation results.  But you should be able to get the same 12 images using the codebase here. We are also working on a huggingface demo using Gradio, and hopefully, we can make it available soon. Hope this helps!

**Comment by Edenzzzz** [comment]
> 

yes, this is indeed very soon.
